### PR TITLE
Enable nothunk tests by setting check-tvar-invariant flag for strict-stm

### PR DESCRIPTION
**Pull request by yogeshsajanikar**
# Description

* Add a flag checkTVarInvariant to ./nix/ouroboros-network.nix
* `haskellPackagesWithTVarCheck` points to `strict-stm` package with "checktvarinvariant" flag enabled.
* `nix-build -A tvar-invariant-checks.tests` runs the tests with "checktvarinvariant" flag turned on
* Github action "Strict NoThunk Test" should run the tests with above flag turned on

Closes input-output-hk/ouroboros-consensus#604 

# Checklist

- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Commits have useful messages
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

### CODE REVIEW

File: .github/workflows/check-invariant-test.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+          repository: input-output-hk/ouroboros-network
+          ref: ${{ github.event.pull_request.head.sha || github.ref }}
+          fetch-depth: 0
+
+      - name: Prepare nix
+        uses: cachix/install-nix-action@v17
+        with:
+          extra_nix_config: |
+            trusted-public-keys = iohk.cachix.org-1:DpRUyj7h7V830dp/i6Nti+NEO2/nhblbov/8MW7Rqoo= hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
+            substituters = https://cache.nixos.org https://hydra.iohk.io https://iohk.cachix.org
```
Comment (amesgen) [inline .github/workflows/check-invariant-test.yml:25]: I think `iohk.cachix.org` is not being used anymore, and `cache.iog.io` is the replacement of `hydra.iohk.io`.

File: .github/workflows/check-invariant-test.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        uses: actions/cache@v2.1.5
+        with:
+          path: |
+            ~/.cabal/packages
+            ~/.cabal/store
+            dist-newstyle
+          key: |
+            cabal-${{ runner.os }}-${{ hashFiles('cabal.project', 'default.nix', 'shell.nix') }}
+          restore-keys: |
+            cabal-${{ runner.os }}-${{ hashFiles('cabal.project', 'default.nix', 'shell.nix') }}
```
Comment (amesgen) [inline .github/workflows/check-invariant-test.yml:36]: As everything is done via Nix, these directories are unused, so this step should be removed.

File: .github/workflows/check-invariant-test.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+      - name: "Build and Test"
+        run: |
+          nix-build -A tvar-invariant-checks.tests
```
Comment (amesgen) [inline .github/workflows/check-invariant-test.yml:40]: WDYT about adding `--keep-going` to ensure that we still run unrelated test suites to completion even if a previous one already failed?

File: .github/workflows/check-invariant-test.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
+name: Strict NoThunk test
+on:
```
Comment (amesgen) [inline .github/workflows/check-invariant-test.yml:2]: Before this is merged, a [`schedule` trigger](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onschedule) should be added here, right?

File: nix/ouroboros-network.nix [hunk 5]
```diff
@@ -43,6 +45,10 @@ let
           doCoverage = config.coverage;
         });
       }
+      {   
+        packages.strict-stm.components.library.configureFlags =
+          lib.mkForce (if checkTVarInvariant then ["-f checktvarinvariant"] else []);
```
Comment (amesgen) [inline nix/ouroboros-network.nix:50]: Slightly more idiomatic:
```suggestion
          lib.mkIf checkTVarInvariant (lib.mkForce ["-f checktvarinvariant"]);
```

**Review by amesgen — CHANGES_REQUESTED** [review]

File: .github/workflows/check-invariant-test.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+jobs:
+  build-test:
+    name: "Build and Test with Strict Check TVar Invariant Check"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2.3.1
+        with:
+          repository: input-output-hk/ouroboros-network
+          ref: ${{ github.event.pull_request.head.sha || github.ref }}
+          fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (yogeshsajanikar) [reply .github/workflows/check-invariant-test.yml:18]: Yes, these are temporary.

File: .github/workflows/check-invariant-test.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      - master
+      - yogeshs/3001-check-tvar-invariant
+
+
+jobs:
+  build-test:
+    name: "Build and Test with Strict Check TVar Invariant Check"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2.3.1
```
[the remark this answers is not in the document]
Comment (yogeshsajanikar) [reply .github/workflows/check-invariant-test.yml:14]: I will check this. Do not know about dependabot

File: .github/workflows/check-invariant-test.yml (hunk 2 above)
Comment (yogeshsajanikar) [reply .github/workflows/check-invariant-test.yml:36]: dist-newstyle cache is probably required. (?)

File: .github/workflows/check-invariant-test.yml (hunk 4 above)
Comment (yogeshsajanikar) [reply .github/workflows/check-invariant-test.yml:2]: Yes!

File: .github/workflows/check-invariant-test.yml (hunk 2 above)
Comment (amesgen) [reply .github/workflows/check-invariant-test.yml:36]: No, everything happens in the nix store, no local files are touched.

File: .github/workflows/check-invariant-test.yml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+          # TODO: Yogesh: Remove this once PR is approved as this action would run on the default branch
+          ref: ${{ github.event.pull_request.head.sha || github.ref }}
+          fetch-depth: 0
+
+      - name: Prepare nix
+        uses: cachix/install-nix-action@v17
+        with:
+          extra_nix_config: |
+            trusted-public-keys =  cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY= hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ=
+            substituters = https://cache.nixos.org https://cache.iog.io https://iohk.cachix.org
```
Comment (amesgen) [inline .github/workflows/check-invariant-test.yml:28]:
```suggestion
            substituters = https://cache.nixos.org https://cache.iog.io
```

**Comment by amesgen** [comment]
Routine cleanup of stale PRs targeting Consensus components (which nowadays live in https://github.com/IntersectMBO/ouroboros-consensus).

No work is lost, see https://stackoverflow.com/a/17954767.

**Pull request closed without merging**